### DISCUSSION TITLE
App boostrap/configuration

### DISCUSSION THREAD

**Discussion by bernatvadell** [directus/directus#6960] [Feature Requests]
Hi guys!

I was thinking that as an alternative to starting the project, it would be interesting to do it from the code of the application itself, as express does. With the aim of making the configuration even more flexible.

I am imagining something like this: 

```ts
import myCustomModule from './my-custom-module';
const builder = new DirectusBuilder();
builder.use(myCustomModule);
const app = builder.build();
app.start();
```

To the extension registry, an array or callback could also be indicated. 

In this way, I think it would be very easy to modularize the app and that installing external extensions would be as simple and configurable as something like this: 

`npm i another-datepicker`

```ts
import anotherDatePicker from 'another-datepicker';
const builder = new DirectusBuilder();
builder.use(anotherDatePicker({ defaultFormat: 'DD-MM-YYY' }));
const app = builder.build();
app.start();
```

I think it is an option that could perfectly coexist with the current startup mode (very simple for new projects)

**Comment by rijkvanzanten** [comment]
Interesting approach!

How would you pass on the configuration that's normally done in env vars? You can theoretically already import just the express app itself, so something like

```js
import createApp from 'directus/dist/app';

const config = {
  dbClient: 'pg',
  dbHost: 'localhost',
  etc: 'etc'
};

const app = await createApp(config);

app.register(myExtension);

app.start();
```

could work 🤔

**Reply by bernatvadell** [reply]
Yes, I don't know if Directus is working on the possibility of implementing multiple data sources in the future. I am imagining something like this: 

_Shared from npm package, ex: @directus/custom-addon_
```ts
import { BuilderStage } from '@directus/app/builder';

export default function myCustomRegister(config){
  return function*(context){
     const startTime = new Date();
     // we can use generator to control lifecycle registration, for example, before load database, after load database, after load items, etc...
     yield BuilderStage.OnDatasourcesLoaded;
     context.addInterface({...});
     yield BuilderStage.OnDirectusInitialized;
     const endTime = new Date();
    const timeElapsed = (endTime - startTime).totalMiliseconds;
  }
}
```
_My app bootstrap_
```ts
import DirectusBuilder from 'directus/dist/builder';
import datasource from 'directus/dist/datasource';
import customAddon from '@directus/custom-addon';

const builder = new DirectusBuilder();

builder.use(datasource({
  name: 'source 1',
  dbClient: 'pg',
  dbHost: 'localhost',
  etc: 'etc'
}));

builder.use(datasource({
  name: 'source 2',
  dbClient: 'pg',
  dbHost: 'localhost',
  etc: 'etc'
}));

builder.use(customAddon({ .... }));

builder.use(directusExpress({ port: 3000 }));

const sdk = await builder.createApp();

const items = await sdk.items.articles.readMany();
```

I have mixed several concepts, including the possibility of decoupling the creation of the directus app from express, perhaps fastify or any other could be implemented.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
